static/app/components/asyncComponentSearchInput.tsx

```diff
@@ -24,129 +46,156 @@ type DefaultProps = {
-    onError: () => void;
-
-    /**
-     * A render-prop child may be passed to handle custom rendering of the input.
-     */
-    children?: (otps: RenderProps) => React.ReactNode;
-  };
-
-type State = {
-  query: string;
-  busy: boolean;
-};
+  api: Client;
+  className?: string;
+  /**
+   * URL to make the search request to
+   */
+  url: string;
+  /**
+   * Updates URL with search query in the URL param: `query`
+   */
+  updateRoute?: boolean;
 
-/**
- * This is a search input that can be easily used in AsyncComponent/Views.
- *
- * It probably doesn't make too much sense outside of an AsyncComponent atm.
- */
-class AsyncComponentSearchInput extends React.Component<Props, State> {
-  static defaultProps: DefaultProps = {
-    placeholder: t('Search...'),
-    debounceWait: 200,
-  };
+  onSearchSubmit?: (query: string, event: React.FormEvent) => void;
+  onSuccess: (data: object, resp: ResponseMeta | undefined) => void;
+  onError: () => void;
+
+  /**
+   * A render-prop child may be passed to handle custom rendering of the input.
+   */
+  children?: (otps: RenderProps) => React.ReactElement;
+}
 
-  state: State = {
+function AsyncComponentSearchInput({
+  placeholder = t('Search...'),
+  debounceWait = 200,
+  api,
+  className,
+  url,
+  updateRoute,
+  onSearchSubmit,
+  onSuccess,
+  onError,
+  router,
+  location,
+  children,
+}: AsyncComponentSearchInputProps): React.ReactElement {
+  const [state, setState] = React.useState<{query: string; busy: boolean}>({
     query: '',
     busy: false,
-  };
-
-  immediateQuery = async (searchQuery: string) => {
-    const {location, api} = this.props;
-    this.setState({busy: true});
-
-    try {
-      const [data, , resp] = await api.requestPromise(`${this.props.url}`, {
-        includeAllArgs: true,
-        method: 'GET',
-        query: {...location.query, query: searchQuery},
-      });
-      // only update data if the request's query matches the current query
-      if (this.state.query === searchQuery) {
-        this.props.onSuccess(data, resp);
+  });
+
+  // We need to use a mutable ref to keep reference to our latest query, else
+  // useCallback scope will always reference the previous value and our condition will always fail
+  const latestQuery = React.useRef<string>(state.query);
+
+  const queryResolver = React.useCallback(
+    async (searchQuery: string) => {
+      latestQuery.current = searchQuery;
+      setState({query: searchQuery, busy: true});
+
+      try {
+        const [data, , resp] = await api.requestPromise(url, {
+          includeAllArgs: true,
+          method: 'GET',
+          query: {...location.query, query: searchQuery},
+        });
+
+        // only update data if the request's query matches the current query
+        if (latestQuery.current === searchQuery) {
+          onSuccess(data, resp);
+        }
+      } catch {
+        // TODO: should this also respect latestQuery === searchQuery?
```

Comment:
I think the onError check should follow the same latestQuery.current === searchQuery otherwise onError may signal a false error state

Reply:
@priscilawebdev any thoughts here :)?

Reply:
I agree with you